### ISSUE TITLE
cy.wait() does not resolve aliases in Cypress 6.0.x

### ISSUE THREAD

**Issue by dominicroessner** [cypress-io/cypress#9526]
### Current behavior

After upgrading from the latest version of v5 to v6.0.0, we are seeing issues with `cy.wait()` not resolving aliases set with `cy.intercept()`. When our app makes the request, Cypress shows that the request matched and the count # under routes shows 1. See screenshot. Additionally when the route is hit, hovering over where the request occurred in the runner shows "<route> ResponseReceived aliased as: pinReport". The subsequent `cy.wait()` command however times out after the default timeout saying that request never occurred. 

Temporarily switching back to `cy.route()` also results in the same issue.

Important to mention also is that this happens about 50% of the time. Sometimes `cy.wait()` knows the request happened and the test passes, other times it times out like in the screenshot below.

![image](https://user-images.githubusercontent.com/45669990/101096823-fd795d80-3574-11eb-838a-3c978795813f.png)

### Desired behavior

`cy.wait()` should resolve the alias when matched by `cy.intercept()`

### Test code to reproduce

```js
  it('should pin a report', () => {
    cy.intercept({
      method: 'POST',
      path: '/api/report_for_report_indices/*/star',
    }).as('pinReport');

    cy.getByTestId('button__pin')
      .first()
      .click({
        force: true
      });

    cy.wait('@pinReport')
      .its('response.statusCode')
      .should('eq', 204);
  });
```

### Versions

Cypress version v6.0.0 and v6.0.1
Worked with Cypress v5.x
Browser: Chrome Version 87.0.4280.67 (Official Build) (x86_64)
Operating System: macOS, also ran reproducible in GitHub Actions on a ubuntu-18 container

**Comment by bahmutov** [comment]
Can I ask you if this is the only test in the spec or are there other tests? Also could you try `cy.wait(100)` between the `cy.intercept` and the `cy.getByTestId` commands? I am thinking maybe there is some leaking of intercepts, and / or race condition between creating an intercept and the command.

**Comment by dominicroessner** [comment]
In this case it is the first of multiple tests in the spec that runs. It is still reproducible if I add a `.only` to the `it` block to run it in isolation. There are other specs that have tests where this happens, similar logic with `cy.intercept` and `cy.wait`. It does not always happen to the first test in the spec where I am seeing this if that helps at all.

I added the `cy.wait(100)` between the `cy.intercept` and `cy.getByTestId` as suggested does not seem to make a difference. Also upped the time a bit, i.e. `cy.wait(1000)`, same results.

One thing that did seem to make a difference was adding a `cy.wait(100)` before AND after the `cy.intercept` like below. This seems to be passing the test every time (adding it ONLY before the `cy.intercept` also did not work).

```
  it('should pin a report', () => {
    cy.wait(100) // added this
    cy.intercept({
      method: 'POST',
      path: '/api/report_for_report_indices/*/star',
    }).as('pinReport');
    cy.wait(100) // and added this

    cy.getByTestId('button__pin')
      .first()
      .click({
        force: true
      });

    cy.wait('@pinReport')
      .its('response.statusCode')
      .should('eq', 204);
  });
```

**Comment by bahmutov** [comment]
@dominicroessner how do you visit the site?

**Comment by bahmutov** [comment]
@dominicroessner does your site use ServiceWorker by any chance?

**Comment by dominicroessner** [comment]
> In this case it is the first of multiple tests in the spec that runs. It is still reproducible if I add a `.only` to the `it` block to run it in isolation. There are other specs that have tests where this happens, similar logic with `cy.intercept` and `cy.wait`. It does not always happen to the first test in the spec where I am seeing this if that helps at all.
> 
> I added the `cy.wait(100)` between the `cy.intercept` and `cy.getByTestId` as suggested but does not seem to make a difference. Also upped the time a bit, i.e. `cy.wait(1000)`, same results.
> 
> One thing that did seem to make a difference was adding a `cy.wait(100)` before AND after the `cy.intercept` like below. This seems to be passing the test every time (adding it ONLY before the `cy.intercept` also did not work).
> 
> ```js
>   it('should pin a report', () => {
>     cy.wait(100) // added this
>     cy.intercept({
>       method: 'POST',
>       path: '/api/report_for_report_indices/*/star',
>     }).as('pinReport');
>     cy.wait(100) // and added this
> 
>     cy.getByTestId('button__pin')
>       .first()
>       .click({
>         force: true
>       });
> 
>     cy.wait('@pinReport')
>       .its('response.statusCode')
>       .should('eq', 204);
>   });
> ```

Quick update, just got another failure for one of our tests in CI even with the `cy.wait(100)` before and after the `cy.intercept`. So although it seems to have reduced the odds of this happening by a good amount, it still does occur.

**Comment by ttomaszewski** [comment]
I'm seeing similar behavior.

**Comment by jarretmoses** [comment]
I believe this is related to the `minimatch` issue reported in other issues (such as https://github.com/cypress-io/cypress/issues/9340). Essentially `.intercept` is not matching glob patterns through string.

**Comment by jennifer-shehane** [comment]
@jarretmoses I don't think this is related to the glob matching, since the issues described above are intermittent and the glob matching issue should be reproducible every test.

**Comment by AElmoznino** [comment]
I'm seeing a similar issue.
```
cy.route('POST', `${Cypress.env('BFF_API_URL')}/_submit`, {}).as('oldFashionedRoute');
      
cy.intercept('POST', `${Cypress.env('BFF_API_URL')}/_submit`, {
  statusCode: 200,
  body: 'it worked!',
}).as('interceptRequest');
```

When I console.log `cy.state('requests')` I can only see `oldFashionedRoute` being aliased, not the `interceptRequest`(it has `alias: undefined`although I can see that request is in the list of requests made).

**Comment by applebya-wgu** [comment]
Possible workaround in the meantime, which fixes the issue for me:

Try moving your `cy.intercept()` into a `beforeEach()` or `before()`, and/or before you call `cy.visit()`. Something about registering the intercept paths early on seems to work.

In my case, the issue itself only appears when I apply an HTTP Interceptor to my Angular 9 application. And it only occurs with GET requests made upon initialization of my application (`ngOnInit()`)

**Comment by baeharam** [comment]
Same issue, but it occurs only at CI. Should be resolved.

**Comment by jennifer-shehane** [comment]
Can anyone provide a reproducible example? Here are some tips for providing a [Short, Self Contained, Correct, Example](http://sscce.org/) and our own [Troubleshooting Cypress](https://on.cypress.io/troubleshooting) guide.

**Comment by flotwig** [comment]
@dominicroessner Try Cypress 6.1.0, there were some bugfixes regarding waiting on `cy.intercept`s in that version.

**Comment by dominicroessner** [comment]
@flotwig I did upgrade to 6.1.0 yesterday and so far everything is running smoothly. I will continue monitoring but it does appear to be resolved.

**Comment by DBLaci** [comment]
Well, something with the url is definetaly changed.

wait for this alias fails, but worked before (with route)

```
    cy.intercept({
      method: 'GET',
      url: '/Admin/api/setOffGetData?item_id=*',
    }).as('setOffGetData')
```

this works:

```
    cy.intercept({
      method: 'GET',
      url: '**/Admin/api/setOffGetData?item_id=*',
    }).as('setOffGetData')
```

But the cypress log shows:

![image](https://user-images.githubusercontent.com/5806644/101665412-3cb01e80-3a4d-11eb-8f83-d790ff0cbb1b.png)

Something is prepended to the url, but it is not visible anywhere. I hope this helps someone.

**Comment by OldSchooled** [comment]
> Well, something with the url is definetaly changed. (cypress 6.0.1 and cypress 6.1.0)
> 
> wait for this alias fails, but worked before (with route)
> 
> ```
>     cy.intercept({
>       method: 'GET',
>       url: '/Admin/api/setOffGetData?item_id=*',
>     }).as('setOffGetData')
> ```
> 
> this works:
> 
> ```
>     cy.intercept({
>       method: 'GET',
>       url: '**/Admin/api/setOffGetData?item_id=*',
>     }).as('setOffGetData')
> ```
> 
> But the cypress log shows:
> 
> ![image](https://user-images.githubusercontent.com/5806644/101665412-3cb01e80-3a4d-11eb-8f83-d790ff0cbb1b.png)
> 
> Something is prepended to the url, but it is not visible anywhere. I hope this helps someone.

`url` now matches the full url, including the base path (hostname). For your first example, you would use `path` instead which excludes the hostname.

**Comment by DW8Reaper** [comment]
I can't tell if this issue is related but I have something very similar happening although for me it happens in 5.6.0 with route2 and also in 6.0.1 and 6.1.0 with intercept (ran into this while attempting to upgrade to 6.1.0 from 5.0.0)

It seems like when routes contain the same parent path that wait picks up the incorrect alias or maybe just cannot find the child route. So if a have a route /some-parent/parent-a and /some-parent/parent-a/some-child the the child routes alias doesn't work. Here are the routes my test registered (and I apologize for blanking out the parent paths I'm limited as to what I can share)
![route-registered-by-test](https://user-images.githubusercontent.com/3433459/102514915-16950a80-405b-11eb-853c-0cec76b85e69.png)

when my test runs though the child route "options" never seems to match instead it looks like it gets the alias from its parent "fullTemplate"

![test-running-routes](https://user-images.githubusercontent.com/3433459/102515007-30365200-405b-11eb-951d-a389b1757101.png)

**Comment by DW8Reaper** [comment]
I created a super simple reproduction on StackBlitz, you will have to run it locally though just do a ```npm start``` then ```npm run test```

https://stackblitz.com/edit/js-h7pfkz?file=package.json

**Comment by jjohnston-roomandboard** [comment]
I tried going to 6.1.0 today and I am seeing this same thing with using intercept. If I try to reuse the same alias then on the second call it times out.

`cy.intercept(`**/api/orders/${orderId}*`).as('orderDetails');
...
cy.wait('@orderDetails')
`
> Timed out retrying: cy.wait() timed out waiting 10000ms for the 2nd request to the route: orderDetails. No request ever occurred

However, if I make the alias name unique then it works.  

`cy.intercept(`**/api/orders/${orderId}*`).as(`orderDetails${orderId}`);
...
cy.wait(`@orderDetails${orderId}`)`

**Comment by jennifer-shehane** [comment]
@DBLaci This is the issue you're seeing, which will be fixed in our next release https://github.com/cypress-io/cypress/pull/14241

@DW8Reaper With your example, the route matches the first cy.intercept() definition for both requests. I don't think this is related to the opened issue - since their alias was matching correctly, but not resolving. 

You'll need to be more specific about the matching of your routes, like using RegExp below:

```js
it("loads requests", () => {
  cy.intercept("GET", /\/api\/parent\/a$/, "parent a data").as("parent-request")
  cy.intercept("GET", "/api/parent/a/child", "child data").as("child-request")
  cy.visit("index.html")
  cy.wait("@parent-request")
  cy.wait("@child-request")
})
```

The original person who opened this issue said it's been resolved since upgrading to 6.1.0. https://github.com/dominicroessner

This issue will be closed to further comment as the exact issue here was resolved. 

If you're experiencing a bug similar to this in Cypress, please [open a new issue](https://github.com/cypress-io/cypress/issues/new/choose) with a fully reproducible example that we can run. There may be a specific edge case with the issue that we need more detail to fix.
